report/metrics.go

```diff
@@ -0,0 +1,268 @@
+// Add method to add values.
+type Metric struct {
+	Samples Samples   `json:"samples"`
+	Min     float64   `json:"min"`
+	Max     float64   `json:"max"`
+	First   time.Time `json:"first"`
+	Last    time.Time `json:"last"`
+}
+
+// Sample is a single datapoint of a metric.
+type Sample struct {
+	Timestamp time.Time `json:"date"`
+	Value     float64   `json:"value"`
+}
+
+// MakeMetric makes a new Metric.
+func MakeMetric() Metric {
+	return Metric{}
+}
+
+// WithFirst returns a fresh copy of m, with First set to t.
+func (m Metric) WithFirst(t time.Time) Metric {
+	m.First = t
+	return m
+}
+
+// Len returns the number of samples in the metric.
+func (m Metric) Len() int {
+	return m.Samples.Size()
+}
+
+// Add adds the sample to the Metric. Add is the only valid way to grow a
+// Metric. Add returns the Metric to enable chaining.
+func (m Metric) Add(t time.Time, v float64) Metric {
+	newSamples := m.Samples
+	popped := Samples{}
+	for !newSamples.IsNil() {
+		head := newSamples.Head()
+		if head.Timestamp.Equal(t) {
+			// The list already has the element.
+			return m
+		}
+		if head.Timestamp.Before(t) {
+			// Reached insertion point.
+			break
+		}
+		newSamples = newSamples.Tail()
+		popped = popped.Cons(head)
+	}
+	newSamples = newSamples.Cons(Sample{Timestamp: t, Value: v})
+	// Re-add any samples after this one.
+	popped.ForEach(func(s Sample) {
+		newSamples = newSamples.Cons(s)
+	})
+	m.Samples = newSamples
+	if v > m.Max {
+		m.Max = v
+	}
+	if v < m.Min {
+		m.Min = v
+	}
+	if m.First.IsZero() || t.Before(m.First) {
+		m.First = t
+	}
+	if m.Last.IsZero() || t.After(m.Last) {
+		m.Last = t
+	}
+	return m
+}
+
+// Merge combines the two Metrics and returns a new result.
+func (m Metric) Merge(other Metric) Metric {
+	other.Samples.ForEach(func(s Sample) {
+		m = m.Add(s.Timestamp, s.Value)
+	})
+	if !other.First.IsZero() && other.First.Before(m.First) {
+		m.First = other.First
+	}
+	if !other.Last.IsZero() && other.Last.After(m.Last) {
+		m.Last = other.Last
+	}
+	if other.Min < m.Min {
+		m.Min = other.Min
+	}
+	if other.Max > m.Max {
+		m.Max = other.Max
+	}
+	return m
+}
+
+// Copy returns a value copy of the Metric. Metric is immutable, so we can skip
+// this.
+func (m Metric) Copy() Metric {
+	return m
+}
+
+// Div returns a new copy of the metric, with each value divided by n.
+func (m Metric) Div(n float64) Metric {
+	oldSamples := m.Samples
+	m.Samples = Samples{}
+	oldSamples.ForEach(func(s Sample) {
+		m = m.Add(s.Timestamp, s.Value/n)
+	})
+	m.Max = m.Max / n
+	m.Min = m.Min / n
+	return m
+}
+
+// LastSample returns the last sample in the metric, or nil if there are no
+// samples.
+func (m Metric) LastSample() *Sample {
+	if m.Samples.IsNil() {
+		return nil
+	}
+	s := m.Samples.Head()
+	return &s
+}
+
+// Samples is an immutable list of timeseries data. We have this to implement
+// proper marshalling for the ps.List, as well as fixing some if ps.List's
+// behaviour.
+type Samples struct {
+	ps.List
+}
```

Comment:
I'm debating where we need the extra nesting (Metric -> Samples -> ps.List) and the ability to deal with Nil.  Mind if I have a play with this code?